examples/features/opentelemetry: demonstrate tracing using OpenTelemetry plugin (#8056)
grpc/grpc-go: examples/features/opentelemetry/README.md

# OpenTelemetry

> This example shows how to configure OpenTelemetry on a client and server, and shows what type of telemetry data it can produce for certain RPC's.

Author: done

> This example shows how to configure OpenTelemetry on a client and server, and shows what type of telemetry data it can produce for certain RPC's.

Author: reached 80 cols, so moved to the next line.

## Try it

> This section shows how to configure OpenTelemetry Metrics and Traces.

Reviewer: we can remove this I think. The first sentence is enough

Author: done

> **1. Run the gRPC Applications:**

Reviewer: same. No need of this

Author: done

# OpenTelemetry

> This example demonstrates how to configure OpenTelemetry on a gRPC client

Reviewer: configure OpenTelemetry Tracing

Author: done

and server, showcasing the telemetry data it produces for RPC interactions.

## Try it

1.  **Run the gRPC Server:**

    ```
    go run server/main.go
    ```

    * This starts the gRPC server, which also exposes a Prometheus

>         metrics endpoint.

Reviewer: ```suggestion
      metrics endpoint.
```

Author: done

2.  **Run the gRPC Client:**

    ```
    go run client/main.go
    ```

    * This starts the gRPC client, which continuously makes RPC calls
        to the server.
    * The client also exposes a Prometheus metrics endpoint.

3.  **View Prometheus Metrics:**

    * **Server Metrics:**

        ```
        curl localhost:9464/metrics
        ```

        * This command retrieves the Prometheus metrics exposed by the

>             gRPC server.

Reviewer: ```suggestion
          gRPC server.
```

Author: done

    * **Client Metrics:**

        ```
        curl localhost:9465/metrics
        ```

        * This command retrieves the Prometheus metrics exposed by the

>             gRPC client.

Reviewer: ```suggestion
          gRPC client.
```

Author: done

## Explanation

> * **Continuous RPC Calls:** The client continuously makes RPC calls to

Reviewer: why this change about metrics? Let's not make too many changes for metrics. Let's keep this PR focused on adding content for tracing only.

Author: done

    the server, generating telemetry data.
* **Prometheus Exporter:** Both the client and server are configured
    with a Prometheus exporter, which listens and provides metrics.
    * The server's Prometheus endpoint defaults to `:9464`.
    * The client's Prometheus endpoint defaults to `:9465`.
* **OpenTelemetry Configuration:** OpenTelemetry is configured on both
    the client and server to capture metrics and traces.
* **Metrics Export:** The OpenTelemetry configuration exports metrics to
    the Prometheus exporter.
* **Viewing Metrics:** By curling the exposed Prometheus ports, you can
    view the metrics recorded by the client and server. These metrics
    provide insights into the performance and behavior of the gRPC calls.

> ## Adding Trace Visualization (Optional)

Reviewer: This section should be after View metrics like View Traces before Explaination.

Author: done

To visualize traces, you'll need to set up an OTLP Collector and a trace
backend like Jaeger. The client and server code are already configured
to send trace data via OTLP.

1.  **Start Jaeger (or another trace backend):**

    * **Local Jaeger (No Docker):**

        ```
        ./jaeger-all-in-one --collector.otlp.enabled
        ```

    * **Docker Jaeger:**

        ```

>         docker run -d -p 16686:16686 -p 14268:14268 \

Reviewer: so, we need to have docker to run this example?

Author: we can do with or without docker

        jaegertracing/all-in-one:latest
        ```

2.  **Start the OpenTelemetry Collector:**

    * **Local Collector:**
        * Install the `otelcol-contrib` binary.

>         * Create a configuration file (e.g., `collector-config.yaml`)

Reviewer: as i mentioned in the meeting. We should not ask the user to create a config.yaml file. We should provide a yaml which should work as is. If user want to modify that, they can

Author: done

            specifying:
            * An OTLP receiver (to receive traces from your client and
                server).
            * A batch processor (recommended for performance).
            * A Jaeger exporter (to send traces to Jaeger).
        * Run the collector: `./otelcol-contrib --config collector-config.yaml`
    * **Docker Collector:**
        * Create a configuration file (e.g., `collector-config.yaml`)
            specifying the same components as above.
        * Run the collector with Docker, mounting your configuration
            file:

            ```

>             docker run -d -p 4317:4317 -p 4318:4318 \

Reviewer: same: can we do it without docker command?

            -v $(pwd)/collector-config.yaml:/etc/otel-collector-config.yaml \
            --name otel-collector otel/opentelemetry-collector-contrib:latest \
            --config=/etc/otel-collector-config.yaml
            ```

3.  **View Traces in Jaeger UI:**

    * Open `http://localhost:16686` in your browser.

>     * Search for traces to visualize the RPC call flow. The traces will

Reviewer: you can mention the name of some of the traces that gRPC emit

Author: done

# OpenTelemetry

This example demonstrates how to configure OpenTelemetry Tracing on a gRPC client
and server, showcasing the telemetry data it produces for RPC interactions.

## Try it

1.  **Run the gRPC Server:**

    ```
    go run server/main.go
    ```

    * This starts the gRPC server, which is configured to send trace data via
    OTLP and expose Prometheus metrics.

2.  **Run the gRPC Client:**

    ```
    go run client/main.go
    ```

    * This starts the gRPC client, which continuously makes RPC calls to the
    server and is configured to send trace data via OTLP and expose

>     Prometheus metrics.

Reviewer: why "expose Prometheus metrics"? Aren't we talking about traces?

Author: done

        server.

    * **Client Metrics:**

        ```
        curl localhost:9465/metrics
        ```

        * This command retrieves the Prometheus metrics exposed by the gRPC
        client.

## View Traces (Optional)

To visualize traces, you'll need to set up an OTLP Collector and a trace
backend like Jaeger. The client and server code are already configured to
send trace data via OTLP.

1.  **Start Jaeger (or another trace backend):**

    * **Local Jaeger (No Docker):**

        ```
        ./jaeger-all-in-one --collector.otlp.enabled
        ```

2.  **Start the OpenTelemetry Collector:**

    * **Local Collector:**
        * Install the `otelcol-contrib` binary.
        * NOTE: To see the builder manifests used for official binaries, check
            [OpenTelemetry Collector Releases](https://github.com/open-telemetry/
            opentelemetry-collector-releases).
        * For the OpenTelemetry Collector Core distribution specifically, see
            [OpenTelemetry Collector Core Releases](https://github.com/open-telemetry/
            opentelemetry-collector-releases/tree/main/distributions/otelcol).

>         * Create a `collector-config.yaml` file with the following content:

Reviewer: we should have the yaml created in the example itself. Not mention in readme. In readme we should just mention the collector will use the created yaml.

Author: done

# OpenTelemetry

This example shows how to configure OpenTelemetry on a client and server, and
shows what type of telemetry data it can produce for certain RPC's.

> ## See Traces

Reviewer: There should be single section See Traces and Metrics. We shouldn't be saying users to start client and server twice each.

> This section shows how to configure OpenTelemetry Tracing and view trace data.

Reviewer: `This section shows how to configure OpenTelemetry Metrics and Traces and visualize them.`

### Try it

> 1.  **Start Jaeger (Only for traces):**

Reviewer: Mention one line about Jaeger from official docs and link the Jaeger readme.

# OpenTelemetry

This example shows how to configure OpenTelemetry on a client and server, and shows what type of telemetry data it can produce for certain RPC's.

## Try it

This section shows how to configure OpenTelemetry Metrics and Traces and visualize them.

**1. Run the gRPC Applications:**

* **Run the gRPC Server:**

    ```
    go run server/main.go
    ```

* **Run the gRPC Client:**

    ```
    go run client/main.go
    ```

**2. View Telemetry Data:**

    ```
    curl localhost:9464/metrics
    ```

    ```
    curl localhost:9465/metrics
    ```

## Explanation

The client continuously makes RPC's to a server. The client and server both expose a Prometheus exporter to listen and provide metrics. This defaults to :9464 for the server and :9465 for the client. The client and server are also configured to output traces directly to their standard output streams using stdouttrace.

> OpenTelemetry is configured on both the client and the server, and exports metrics to the Prometheus exporter. The exporter exposes metrics on the Prometheus ports described above. OpenTelemetry also exports traces using the stdouttrace exporter. These traces are printed as structured data to the console output of both the client and the server.

Author: done

# OpenTelemetry

This example shows how to configure OpenTelemetry on a client and server, and shows
what type of telemetry data it can produce for certain RPC's.
This example demonstrates how to configure OpenTelemetry Tracing on a gRPC client
and server, showcasing the trace data it produces for RPC interactions.

## Try it

```
go run server/main.go
```
1.  **Run the gRPC Server:**

```
go run client/main.go
```
    ```
    go run server/main.go
    ```

```

> curl localhost:9464/metrics

Reviewer: We can't remove these because they are for metrics. 

Create separate sections "See metrics" and "See traces". "See metrics" can have this part and "See traces" should have collector part.

    server and is configured to send trace data via OTLP.

3. **Start the OpenTelemetry Collector:**

>     * **Local Collector:**

Reviewer: Does this need to be running before client and server? If yes, we need to move this up and and mention in brackets (Only for traces)

        * Install the `otelcol-contrib` binary.
        * NOTE: To see the builder manifests used for official binaries, check 
        [OpenTelemetry Collector Releases]
        (https://github.com/open-telemetry/opentelemetry-collector-releases).
        * For the OpenTelemetry Collector Core distribution specifically, see 
        [OpenTelemetry Collector Core Releases]
        (https://github.com/open-telemetry/opentelemetry-collector-releases/tree/main/distributions/otelcol).
        * The collector will use the `collector-config.yaml` file present in the same directory.
        * **Execute the collector:**
            ```bash
            /path/to/otelcol-contrib --config=collector-config.yaml
            ```
            (Replace `/path/to/otelcol-contrib` with the actual path to your `otelcol-contrib` binary.)

> 4.  **View Traces in the Collector Output:**

Reviewer: Can't we do `curl` similar to metrics?

        * NOTE: To see the builder manifests used for official binaries, check 
        [OpenTelemetry Collector Releases]
        (https://github.com/open-telemetry/opentelemetry-collector-releases).
        * For the OpenTelemetry Collector Core distribution specifically, see 
        [OpenTelemetry Collector Core Releases]
        (https://github.com/open-telemetry/opentelemetry-collector-releases/tree/main/distributions/otelcol).
        * The collector will use the `collector-config.yaml` file present in the same directory.
        * **Execute the collector:**
            ```bash
            /path/to/otelcol-contrib --config=collector-config.yaml
            ```
            (Replace `/path/to/otelcol-contrib` with the actual path to your `otelcol-contrib` binary.)

OpenTelemetry is configured on both the client and the server, and exports to
the Prometheus exporter. The exporter exposes metrics on the Prometheus ports
described above.
4.  **View Traces in the Collector Output:**

    * The OpenTelemetry Collector, using the `debug` exporter, will print trace
    data to the console.
    * Monitor the collector's console output to see the trace information.

## Explanation

> Curling to the exposed Prometheus ports outputs the metrics recorded on the

Reviewer: Same. We can't remove this. As this is for metrics.

# OpenTelemetry

This example shows how to configure OpenTelemetry on a client and server, and shows what type of telemetry data it can produce for certain RPC's.

## Try it

This section shows how to configure OpenTelemetry Metrics and Traces and visualize them.

**1. Start Jaeger (for Traces):**

Jaeger is an open-source, end-to-end distributed tracing system. For more detailed information, refer to the [Jaeger README]
(https://github.com/jaegertracing/jaeger/blob/main/README.md).

* **Server Jaeger Instance:**

    ```
    JAEGER_QUERY_PORT=16687 \
    JAEGER_ADMIN_HTTP_PORT=:14270 \
    JAEGER_COLLECTOR_OTLP_ENABLED=true \

>     ./jaeger-all-in-one \

Reviewer: I believe we need to install jaegar and point in our local command line? Can you please point to installation instruction above?

      --collector.grpc-server.host-port=:14252 \
      --collector.http-server.host-port=:14272 &
    ```
* **Client Jaeger Instance:**

    ```
    JAEGER_QUERY_PORT=16686 \
    JAEGER_ADMIN_HTTP_PORT=:14271 \
    JAEGER_COLLECTOR_OTLP_ENABLED=true \
    ./jaeger-all-in-one \
      --query.http-server.host-port=:16686 \
      --admin.http.host-port=:14271 \
      --collector.otlp.grpc.host-port=:4321 \
      --collector.otlp.http.host-port=:4319 \
      --query.grpc-server.host-port=:16684 &
    ```

**2. Run the gRPC Applications:**

* **Run the gRPC Server:**

    ```
    go run server/main.go
    ```

* **Run the gRPC Client:**

    ```
    go run client/main.go
    ```

**3. View Telemetry Data:**

* **View Metrics:**

    ```
    curl localhost:9464/metrics
    ```

    ```
    curl localhost:9465/metrics
    ```

* **View Traces:**

    ```
    curl -X GET "http://localhost:16687/api/traces?service=grpc-server"
    ```

    ```
    curl -X GET "http://localhost:16686/api/traces?service=grpc-client"
    ```

* **View Traces (Jaeger UI):**

    ```
    http://localhost:16687
    ```

    ```
    http://localhost:16686
    ```

## Explanation

> The client continuously makes RPC's to a server. The client and server both expose a Prometheus exporter to listen and provide metrics. This defaults to :9464 for the server and :9465 for the client. OpenTelemetry is also configured to capture and export trace data for distributed tracing. This can be viewed on the Jaeger UI which defaults to :16687 for the server and :16686 for the client.

Reviewer: ```suggestion
The client continuously makes RPC's to a server. The client and server both expose a Prometheus exporter to listen and provide metrics. This defaults to :9464 for the server and :9465 for the client. The client and server also expose OTLP exporter to capture and export traces. The traces can be viewed on the Jaeger UI which defaults to :16687 for the server and :16686 for the client.
```

Author: done

> OpenTelemetry is configured on both the client and the server, and exports metrics to the Prometheus exporter. The exporter exposes metrics on the Prometheus ports described above. OpenTelemetry also exports trace data using the OpenTelemetry Protocol (OTLP). This trace data can be visualized using a distributed tracing backend like Jaeger.

Reviewer: ```suggestion
OpenTelemetry is configured on both the client and the server, and exports metrics to the Prometheus exporter. The exporter exposes metrics on the Prometheus ports described above. OpenTelemetry also exports traces using the OpenTelemetry Protocol (OTLP). These traces can be visualized using a distributed tracing backend like Jaeger.
```

Author: done

# OpenTelemetry

> This example demonstrates how to configure OpenTelemetry Tracing on a gRPC

Reviewer: No need to change this. The existing language is generic enough. This example is going to show both tracing and metrics. Not just tracing.

# OpenTelemetry

This example shows how to configure OpenTelemetry on a client and server, and

> shows what type of telemetry data it can produce for certain RPC's.

Reviewer 2: ```suggestion
shows what type of telemetry data it can produce for certain RPCs.
```

Author: done

## Try it

```
go run server/main.go
```

```
go run client/main.go
```

```
curl localhost:9464/metrics
curl localhost:9465/metrics
```

## Explanation

> The client continuously makes RPC's to a server. The client and server both

Reviewer 2: ```suggestion
The client continuously makes RPCs to a server. The client and server both
```

Author: done

go run client/main.go
```

> **2. View Telemetry Data:**

Reviewer: no need of this

```
curl localhost:9464/metrics
curl localhost:9465/metrics
```

## Explanation

> The client continuously makes RPC's to a server. The client and server both expose a Prometheus exporter to listen and provide metrics. This defaults to :9464 for the server and :9465 for the client. The client and server are also configured to output traces directly to their standard output streams using stdouttrace.

Author: done
